03-attribute-operations.Rmd

```diff
@@ -185,69 +198,34 @@ knitr::kable(data_frame(Symbol = operators, Name = operators_exp))
-
-```{r}
-world7 = world %>%
-  select(name_long, continent)
-```
-
-Note that this can also be written without the pipe operator because, in the above code, the `world` object is simply 'piped' into the first argument of `select()`.
-The equivalent **dplyr** code without the pipe operator is:
+For example, let us first take the `world` dataset, then let us select the two columns named `name_long` and `continent`, and then we just would like to return the first five rows.
 
 ```{r}
-world8 = select(world, name_long, continent)
-```
-
-`pull()` retrieves a single variable by name or position and returns a vector:
-
-```{r, eval=FALSE}
-world %>% 
-  pull(name_long)
-```
-
-The pipe operator can be used for many data processing tasks with attribute data.
-
-```{r}
-# 1,000,000,000 could be expressed as 1e9 in the scientific notation 
 world %>%
-  filter(pop > 1e9) 
-```
-
-This is equivalent to the following base R code (not run to preserve the NAs):^[[Note](https://github.com/Robinlovelace/geocompr/issues/28) NAs do not work for subsetting by inequalities in base R, hence conversion of NAs to 0s in this version)]
-
-```{r, eval=FALSE}
-# subsetting simple feature rows by values
-world$pop[is.na(world$pop)] = 0 # set NAs to 0
-world_few_rows = world[world$pop > 1e9,]
+  select(name_long, continent) %>%
+  slice(1:5)
 ```
 
-The ` %>% ` operator works the best for combining many operations.
-For example, we want to (1) rename the `name_long` column into a `name` column, (2) picks only `name`, `subregion` and `gdpPercap` and (3) subset countries from "Eastern Asia" with gross domestic product per capita larger than 30,000$:
-
-```{r}
-world %>% 
-  select(name = name_long, subregion, gdpPercap) %>% 
-  filter(subregion == "Eastern Asia", gdpPercap > 30000)
-```
+The pipe operator supports an intuitive data analysis workflow (first do that, then do that, then...). 
+It also lets you read this workflow from left to right, and avoids less easier to read nesting, i.e., to read workflows from the inside to the outside as is commonly the case when using base R.
+Another advantage over the nesting approach is that you can easily comment out certain parts of a pipe.
+**dplyr** works especially well with the pipe operator because its fundamental functions (or 'verbs', like `select()`) expect a data frame object as input and also return one.^[If you want **dplyr** to return a vector, use `pull`.]
 
 ## Attribute data aggregation 
 
 <!-- https://github.com/ropenscilabs/skimr ?? -->
 
-As demonstrated in chapter \@ref(spatial-class), `summary()` provides a quick summary of the spatial and non-spatial components of spatial objects.
-Enter the following command to for an overview of the `world` object and all its variables (result not shown):
+<!-- As demonstrated in chapter \@ref(spatial-class), `summary()` provides a quick summary of the spatial and non-spatial components of spatial objects.
```

Comment:
The comparison is a bit unfair. The summary function is more generic in nature and can be applied to a multitude of classes for summary statistics.
dplyr::summarize is basically an aggregation function, hence, a comparison with `tapply`, `aggregate` or `by` would be fairer. Ok, you can also use dplyr::summarize for summary statistics but so you can with `aggregate`, etc. So I suggest to either drop the summary-comparison or to compare with a base R aggregate function.